### ISSUE TITLE
Non blocking IO problems

### ISSUE THREAD

**Issue by zeus-cht** [pmodels/mpich#5920]
About the non blocking write operation, which is MPI_File_iwrite, I wonder how to use this function correctly? In my expection, when I call MPI_File_iwrite, it should immediately return, and the process can do some computing. In this case, the IO operation and the computing operation are overlapped, and when the computing ends, I call MPI_wait to finish the IO operation. If the compute time is long enough, the MPI_wait should return immediately since the IO oeration is complete during the computing time. However, the truth is that MPI_File_iwrite immediately returned while MPI_wait cost a long time(which is just the time the write operation needs). It indicates that the IO and compute are not overlapped. And we I add a MPI_Test between MPI_File_iwrite and MPI_wait, the MPI_Test cost a long time(just the same as the MPI_wait before), and MPI_wait returned immediately. 
`init file size = 0B `
`open time: 0.158`
`after cpmpute (before wait) file size = 1072128008B`
`after wait file size = 1073741824B`
`total time: 6.203`
`compute time: 4.527`
`file_iwrite time: 0.02775`
`wait time: 1.45986`
environment: centOS7, mpich 3.4, daos file system

**Comment by zeus-cht** [comment]
The code is like this:
```c
    MPI_FIle_iwrite();
    //computation, time is long enough to cover the write operation)
    MPI_wait();  //to check if the MPI_FIle_iwrite is complete
```

**Comment by roblatham00** [comment]
Your expectations about non-blocking I/O are reasonable, but that's not quite how non-blocking I/O operations are implemented in MPICH.

We have two kinds of asynchronous progress here:
-   the I/O operation makes progress when the I/O library works on the operation.  In linux libc, the AIO routines spawn a thread and work on I/O in the background.  DAOS relies on `daos_event_test` to make progress (see https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/ad_daos/ad_daos_io.c#L166 ) .  
- MPICH makes progress on a request object when the progress engine can run on it.  MPICH's progress engine "turns" when MPI_TEST, MPI_WAIT, etc are called.

(Not what you asked, but I'll add it here:  If the underlying file system has no asynchronous routines, ROMIO just does all the I/O up front and hands the user a completed request object.  )

What you are observing is an immediate return, but DAOS cannot work on the I/O request until something calls `daos_event_test`, and only MPI test/wait routines will trigger that function call.

If you have a sophsticated computation,  I would suggest sneaking a MPI_TEST call in between stage of that computation. 

```
   while (!converged) {
       MPI_TEST()
       newton_aproximate(x, y, z, &converged)
   }
```
then once your compute is done your i/o will be waiting for you.

**Comment by zeus-cht** [comment]
Thanks！ @roblatham00 But the MPI_Test() will also  block when using DAOS file system. So I run the same test using Lustre file system, and the results met my expectations：
```
total time: 3.405800 
compute time: 3.000076 
MPI_File_iwrite time: 0.000146 
MPI_Wait time: 0.000123
```
Thus, I think there are some problems in my way of using DAOS.

**Comment by mchaarawi** [comment]
yes indeed currently MPI_test correponds to a daos_event_test with WAIT for the function to complete.
on my side, which i implemented this, i observed that there was a hang if i added a NO_WAIT test:

**Comment by mchaarawi** [comment]
so i think ideally you want to apply this patch::
`diff --git a/src/mpi/romio/adio/ad_daos/ad_daos_io.c b/src/mpi/romio/adio/ad_daos/ad_daos_io.c
index 91ed858..10c56c4 100644
--- a/src/mpi/romio/adio/ad_daos/ad_daos_io.c
+++ b/src/mpi/romio/adio/ad_daos/ad_daos_io.c
@@ -170,14 +170,14 @@ int ADIOI_DAOS_aio_poll_fn(void *extra_state, MPI_Status * status)
     bool flag;
 
     /* MSC - MPICH hangs if we just test with NOWAIT.. */
-    ret = daos_event_test(&aio_req->daos_event, DAOS_EQ_WAIT, &flag);
+    ret = daos_event_test(&aio_req->daos_event, DAOS_EQ_NOWAIT, &flag);
     if (ret != 0)
         return MPI_UNDEFINED;
 
     if (flag)
         MPI_Grequest_complete(aio_req->req);
     else
-        return MPI_UNDEFINED;
+        return ret;
 
     if (aio_req->daos_event.ev_error != 0)
         ret = ADIOI_DAOS_err("ADIOI_DAOS_aio_poll_fn", "DAOS Event Error", __LINE__, ret);
`
but when i tested this (granted a long time ago) this was just causing a hang somewhere in the mpich stack.

**Comment by mchaarawi** [comment]
> > when i tested this (granted a long time ago) this was just causing a hang somewhere in the mpich stack.
> 
> Sounds like this is a bug that we need to track down. @mchaarawi Can you find out where it is hanging? @roblatham00 Is there a machine at Argonne that we can debug DAOS?

i'll try to do that once i get other urgent things off my plate.
as for a machine at ANL with DAOS servers, yes you have a system called Presque. you can contact Kevin H. for access to that i believe.

**Comment by mchaarawi** [comment]
it sounds like this issue was fixed at some point, as my tests with nowait pass now. ive created a PR for this fix:
https://github.com/pmodels/mpich/pull/5941

**Comment by mchaarawi** [comment]
@roblatham00 @hzhou could you please inspect the patch? i don't have permission to add reviewers there. thanks
